### PR TITLE
Chain mrjar source sets

### PR DESCRIPTION
**Pull request by rjernst**
When building MR jars we may have code for several java versions. For example, we could have main21 and main22. This commit adjust the sourcesets to chain together, so that eg main22 can utilize code from main21 (which in turn can already use code from main).

### CODE REVIEW

**Comment by rjernst** [comment]
cc @ChrisHegarty

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java [hunk 1]
```diff
@@ -56,24 +59,38 @@ public void apply(Project project) {
...
+    private void addMrjarSourceset(
+        Project project,
+        JavaPluginExtension javaExtension,
+        String sourcesetName,
+        String parentSourceSetName,
+        int javaVersion
+    ) {
         SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
-        GradleUtils.extendSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME, sourcesetName);
+        GradleUtils.extendSourceSet(project, parentSourceSetName, sourcesetName);
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:93]: I don't think this is transitive. That is, if you have `main`, `22` and `23` then `23` will only see output from `22` but not `main` which might cause issues. I believe you'll need to explicitly call this for all earlier versions.

**Review by mark-vieira — CHANGES_REQUESTED** [review]

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java (hunk 1 above)
Comment (rjernst) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:93]: The docs for extendsFrom say:

> The artifacts of the super configurations are also available in this configuration.

I also setup a dummy test with additional source sets, and it seemed to work (main classes were available in main22).

Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:93]: Indeed `configuration.extendsFrom` is transitive but we are also adding the parent source set output to the classpath but we don't recurse that bit.

> I also setup a dummy test with additional source sets, and it seemed to work (main classes were available in main22).

Was there a `main21` in this scenario?

Comment (rjernst) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:93]: There was. However, I recreated it and found you are right. What's confusing is IntelliJ thinks it's ok, but then at runtime it is not.

Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:93]: Never trust IntelliJ 😉

Comment (rjernst) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java]: This should be addressed in [b109491](https://github.com/elastic/elasticsearch/pull/106092/commits/b10949150176e3ab1cfe299f470ac9fe7cdc9d3d)

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Review by mark-vieira — APPROVED** [review]

**Pull request merged**